### PR TITLE
fix(oauth2) ownership check

### PR DESCRIPTION
**Pull request by thefosk**
**Only merge this feature in a major release like 0.10.**
### Summary

OAuth 2 fix that checks if a code or a token effectively belong to the API that has been used when generating them.

⚠️ This security fix will **permanently delete** all the authorization codes and tokens currently stored in Kong, and the clients will have to require a new token. The token invalidation is required to preserve the security of the system.
### Full changelog
- Implemented security check that verifies if an authorization code, token or refresh token, effectively belong to the API that generated them.

### CODE REVIEW

**Comment by thefosk** [comment]
TODO: This plugin should also introduce an option to actually share the codes/tokens across multiple APIs, since some users liked the old behavior.

**Comment by djMax** [comment]
Why would we truncate the table of codes? Why not just make those codes behave as they do now and add the column to allow api-specific behavior? And allow that to be a configured behavior?

**Comment by thefosk** [comment]
@djMax because now both authorization codes and tokens need to have a mandatory relationship with the API object. The biggest pain for this patch won't be losing the authorization codes (which expire after 5 minutes by default), but losing the tokens which cannot be migrated and associated to any specific API because of how the system is working as of today.

**Comment by thefosk** [comment]
@deffer @djMax 

> use asterisk character to indicate that token/code is global scope

This will break the database relationship with the API table. I will try to investigate a way of doing it before merging, but I cannot guarantee I will ultimately find a clean way of implementing it.

> And have it configurable somewhere whether we want tokens API scope or global scope.

Some users liked the idea to share codes/tokens across multiple APIs (the old behavior), but I don't think this should be a feature of the OAuth2 plugin, but a core feature of how Consumers work.

**Comment by djMax** [comment]
I do not believe you should break existing behavior like this... OAuth2 is a big mess at the moment in Kong, I'm all for rethinking it more broadly, but piecemeal breakage seems like a disaster...

**Comment by djMax** [comment]
(and can we not use NULL to indicate global applicability?)

**Comment by thefosk** [comment]
I kept this PR in standby for now until I finish some other priorities, and then I will resume the conversation when I am back at it. If there is a clean way to keep retro-compatibility I will implement it. NULL could be an option.

**Comment by thefosk** [comment]
The plugin configuration has a new property called `global_credentials = true/false`.

Any credential created for an API that has the plugin configured with `config.global_credentials=true` can be used to consume any other API that has the OAuth2 plugin configured with `config.global_credentials=true` too.

By default `config.global_credentials=false` but in the migration I set it as `true` for all the existing plugin configurations, so that the old behavior doesn't change. 

But if you update an existing plugin with `config.global_credentials=false`, then all the tokens have to be regenerated for that specific API.

File: kong/plugins/oauth2/migrations/cassandra.lua [hunk 1]
```diff
@@ -119,5 +119,34 @@ return {
...
+    ]]
+  },
+  {
+    name = "2016-12-15-set_global_credentials",
+    up = function(_, _, dao)
+      local rows, err = dao.plugins:find_all({name = "oauth2"})
+      if err then return err end
+      
+      for i = 1, #rows do
+        local row = rows[i]
```
Comment (Tieske) [inline kong/plugins/oauth2/migrations/cassandra.lua]: why not `for _, row in ipairs(rows) do` ?

File: kong/plugins/oauth2/migrations/postgres.lua [hunk 2]
```diff
@@ -126,5 +126,20 @@ return {
...
     end
+  },
+  {
+    name = "2016-12-15-set_global_credentials",
+    up = function(_, _, dao)
+      local rows, err = dao.plugins:find_all({name = "oauth2"})
+      if err then return err end
+      
+      for i = 1, #rows do
+        local row = rows[i]
```
Comment (Tieske) [inline kong/plugins/oauth2/migrations/postgres.lua]: same remark as with the cassandra migration

**Comment by mdaguete** [comment]
Hi,

Is this patch applied in 0.11 ? 

There's no reference to `config.global_credentials` in plugin docs.

Thanks. Regards.

**Comment by thefosk** [comment]
@mdaguete yes, but the doc is outdated. I am pushing a new version of the doc as we speak.

**Pull request merged**